### PR TITLE
Reset processors after each batch to be able to re-use

### PR DESCRIPTION
**Pull request by aerdem4**

### CODE REVIEW

File: logits_processor_zoo/transformers/generation_length.py [hunk 1]
```diff
@@ -38,17 +39,20 @@ class GenLengthLogitsProcessor:
...
         self.boost_factor = boost_factor
         self.p = p
-        self.token_count = 0
         self.full_stop_token = text_to_token(tokenizer, "It is a sentence.", last=True)
         self.new_line_token = text_to_token(tokenizer, "It is a new line\n", last=True)
         self.complete_sentences = complete_sentences
 
-    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.Tensor:
+    def _reset(self):
+        self.token_count = 0
```
Comment (maxjeblick) [inline logits_processor_zoo/transformers/generation_length.py:53]: I'd either define `self.token_count = 0` in `init` (same for other processors) or call `self._reset() ` in init to achieve the same. Setting the attributes explicitly is probably the easiest solution (as one wouldn't need to worry/keep in mind to call _reset).


Reason: In some cases (e.g. tests), `._process` might be called instead of `__call__` which would then yield to an error.

File: logits_processor_zoo/transformers/base.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class BaseLogitsProcessor:
+    def __init__(self):
+        self.input_len = None
+
+    def _reset(self):
+        pass
+
+    def _reset_if_new_batch(self, input_ids: torch.LongTensor):
```
Comment (maxjeblick) [inline logits_processor_zoo/transformers/base.py:28]: Could you add a docstring explaining this logic?

File: logits_processor_zoo/vllm/trigger_phrase.py [hunk 3]
```diff
@@ -40,14 +40,15 @@ def __init__(self, phrase: str, trigger_token_phrase: str, tokenizer: PreTrained
...
+    def _reset(self):
         self.index = -1
-        self.trigger_count = trigger_count
+        self.trigger_count = self.initial_trigger_count
 
     def __call__(self, prompt_tokens_ids: List[int], past_token_ids: List[int], scores: torch.Tensor) -> torch.Tensor:
-        if not past_token_ids:
-            self.index = -1
-            self.trigger_count = self.initial_trigger_count
+        if not past_token_ids:  # new generation
```
Comment (maxjeblick) [inline logits_processor_zoo/vllm/trigger_phrase.py:50]: Why aren't you using a similar base class for vllm?

**Review by maxjeblick — CHANGES_REQUESTED** [review]
Thanks a lot for the PR, looks good overall!
I left a few comments, most of them should be easy to address.
Apart from that, it could make sense to introduce a `BaseLogitsProcessor` for vllm, as well, to have a unified interface.

File: logits_processor_zoo/vllm/trigger_phrase.py (hunk 3 above)
Comment (aerdem4) [reply logits_processor_zoo/vllm/trigger_phrase.py:50]: vllm provides past_token_ids. If it is an empty list or None, we can be sure that it is a new generation. In transformers, we actually make an hack to figure out new generations (new batches), which dont guarantee 100% correctness.

File: README.md [hunk 4]
```diff
@@ -78,4 +78,12 @@ I am getting a lot of calls during the day. What is more important for me to con
 2. Operating System
 3. Battery
 ```
-The goal is to make LLM generate "3" as an answer.
\ No newline at end of file
+The goal is to make LLM generate "3" as an answer.
+
+### TriggerPhraseLogitsProcessor
+A logits processor which triggers phrases when it encounters given token.
```
[the remark this answers is not in the document]
Comment (aerdem4) [reply README.md]: Thanks, fixed it.

File: logits_processor_zoo/transformers/base.py (hunk 2 above)
Comment (aerdem4) [reply logits_processor_zoo/transformers/base.py:28]: I didn't add a docstring because it is a private function that I dont want to expose to users but I added comments to make it more clear if it is okay?

File: logits_processor_zoo/transformers/generation_length.py (hunk 1 above)
Comment (aerdem4) [reply logits_processor_zoo/transformers/generation_length.py:53]: Since they import the base class, _reset is guaranteed to be called on behalf of the user.

We actually dont want users to call _process, therefore I named it with underscore to make it a private function.

File: logits_processor_zoo/transformers/base.py (hunk 2 above)
Comment (maxjeblick) [reply logits_processor_zoo/transformers/base.py:28]: Feel free to add a comment, not curcial.

**Review by maxjeblick — APPROVED** [review]
PR looks good, thanks for adding reset functionality!
